### PR TITLE
kubernetes api deployment can setup openID connect

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: configure k8s API server settings in clusters to accept OpenID. The OIDC settings are optional and can be enabled by passing additional command line flags.

```release-note
Kubernetes API servers can now be used with OpenID authentication
[ACTION REQUIRED] to enable the OpenID for kubernetes API server the users must set `-feature-gates=OpenIDConnectTokens=true` and provide `-oidc-issuer-url`, `-oidc-issuer-client-id` when running the controller.
```

### CODE REVIEW

File: api/cmd/kubermatic-controller-manager/options.go
Comment (p0lyn0mial) [inline api/cmd/kubermatic-controller-manager/options.go]: I think that we would like to pass one more flag `--oidc-ca-file` as described in the proposal https://github.com/kubermatic/kubermatic/blob/master/docs/proposals/user-cluster-authentication-via-dex.md

File: api/cmd/kubermatic-controller-manager/options.go
Comment (p0lyn0mial) [inline api/cmd/kubermatic-controller-manager/options.go]: should we validate if `oidcURL` holds a valid URL ?  (using https://golang.org/pkg/net/url/#Parse ?)

File: api/cmd/kubermatic-controller-manager/features.go
Comment (p0lyn0mial) [inline api/cmd/kubermatic-controller-manager/features.go]: nit: it actually activates the auth plugin, how about changing this field to `OpenIDAuthPlugin` ?

File: api/pkg/resources/data.go
Comment (p0lyn0mial) [inline api/pkg/resources/data.go]: If you decide to change the name of `OpenIDConnectTokens` then I'd change the name of this function as well, `OIDCAuthPluginEnabled` ?

File: api/pkg/resources/data.go
Comment (p0lyn0mial) [inline api/pkg/resources/data.go]: instead of defining  a new variable we could determine if the functionality was enabled by `return len(d.oidcURL) > 0 && len(d.oidcIssuerClientID) > 0)`

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/cmd/kubermatic-controller-manager/options.go
Comment (p0lyn0mial) [reply api/cmd/kubermatic-controller-manager/options.go]: you could test it by passing letsencrypt's CA - https://letsencrypt.org/certificates/
I think that dex's cert is letsencrypt.

**Comment by p0lyn0mial** [comment]
@alvaroaleman, @mrIncompetent this is important and has a high priority, so please PTAL.

File: api/cmd/kubermatic-controller-manager/main.go
Comment (alvaroaleman) [inline api/cmd/kubermatic-controller-manager/main.go]: IMHO we should treat the `options` as immutable after we validated them, so can we move this before the validation?

File: api/cmd/kubermatic-controller-manager/options.go
Comment (alvaroaleman) [inline api/cmd/kubermatic-controller-manager/options.go]: To me it feels more natural to follow the naming of the `kube-apiserver` which names this option `--oidc-issuer-url`, is there a reason not to do that here?

File: api/pkg/resources/apiserver/deployment.go
Comment (alvaroaleman) [inline api/pkg/resources/apiserver/deployment.go]: We probably do not want to hardcode this to `dev.kubermatic.io` ;)

Comment (zreigz) [reply api/pkg/resources/apiserver/deployment.go]: done

File: api/cmd/kubermatic-controller-manager/options.go
Comment (zreigz) [reply api/cmd/kubermatic-controller-manager/options.go]: done

File: api/cmd/kubermatic-controller-manager/main.go
Comment (zreigz) [reply api/cmd/kubermatic-controller-manager/main.go]: this part was removed

File: api/pkg/resources/data.go
Comment (zreigz) [reply api/pkg/resources/data.go]: done

File: api/pkg/resources/data.go
Comment (zreigz) [reply api/pkg/resources/data.go]: done

File: api/cmd/kubermatic-controller-manager/features.go
Comment (zreigz) [reply api/cmd/kubermatic-controller-manager/features.go]: done

File: api/cmd/kubermatic-controller-manager/options.go
Comment (zreigz) [reply api/cmd/kubermatic-controller-manager/options.go]: done

File: api/cmd/kubermatic-controller-manager/options.go
Comment (zreigz) [reply api/cmd/kubermatic-controller-manager/options.go]: the flag is added

File: api/cmd/kubermatic-controller-manager/options.go
Comment (zreigz) [inline api/cmd/kubermatic-controller-manager/options.go:53]: I think it should be a boolean flag because this file will be from a secret?

Comment (p0lyn0mial) [reply api/cmd/kubermatic-controller-manager/options.go:53]: I think so, we could rename it to `skip-oidc-tls-verify` or `insecure-skip-oidc-tls-verify` ? It could be `on` by default.

**Review by p0lyn0mial — APPROVED** [review]
/lgtm
@alvaroaleman please PTAL

@alvaroaleman @zreigz I think we could merge this pull and add TLS verification in a separate pull. I think that the changes we made in this pull don't break anything. WDYT ?

File: api/cmd/kubermatic-controller-manager/options.go
Comment (p0lyn0mial) [reply api/cmd/kubermatic-controller-manager/options.go:53]: we could add TLS verification in a separate pull.

**Comment by zreigz** [comment]
Fine for me. I have been working with new PR and I will add TLS verification + new secret with Dex CA

**Review by alvaroaleman — APPROVED** [review]
lgtm, great work. I am afraid thought we have to wait for @mrIncompetent to approve as he is a codeowner here and also the only one with the required permissions to ignore that check.

Looking forward to have Prow and being able to define multiple codeowners

**Comment by alvaroaleman** [comment]
@p0lyn0mial  doing tls verification in a separate pull sounds good

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**
